src/Microsoft.Extensions.Hosting/Internal/Host.cs

```diff
@@ -0,0 +1,98 @@
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Extensions.Hosting.Internal
+{
+    internal class Host : IHost
+    {
+        private ILogger<Host> _logger;
+        private IHostLifetime _hostLifetime;
+        private ApplicationLifetime _applicationLifetime;
+        private IEnumerable<IHostedService> _hostedServices;
+
+        internal Host(IServiceProvider services)
+        {
+            Services = services ?? throw new ArgumentNullException(nameof(services));
+            _applicationLifetime = Services.GetRequiredService<IApplicationLifetime>() as ApplicationLifetime;
+            _logger = Services.GetRequiredService<ILogger<Host>>();
+            _hostLifetime = Services.GetService<IHostLifetime>() ?? new NoOpHostLifetime();
+        }
+
+        public IServiceProvider Services { get; }
+
+        public async Task StartAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            _logger.Starting();
+
+            var delayStart = new TaskCompletionSource<object>();
+            cancellationToken.Register(obj => ((TaskCompletionSource<object>)obj).TrySetCanceled(), delayStart);
+            _hostLifetime.OnStarted(obj => ((TaskCompletionSource<object>)obj).TrySetResult(null), delayStart);
+            _hostLifetime.OnStopping(obj => (obj as IApplicationLifetime)?.StopApplication(), _applicationLifetime);
+
+            await delayStart.Task;
+
+            _hostedServices = Services.GetService<IEnumerable<IHostedService>>();
+
+            foreach (var hostedService in _hostedServices)
+            {
+                // Fire IHostedService.Start
+                await hostedService.StartAsync(cancellationToken).ConfigureAwait(false);
+            }
+
+            // Fire IApplicationLifetime.Started
+            _applicationLifetime?.NotifyStarted();
+
+            _logger.Started();
+        }
+
+        public async Task StopAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            _logger.Stopping();
+            
+            // Trigger IApplicationLifetime.ApplicationStopping
+            _applicationLifetime?.StopApplication();
+
+            IList<Exception> exceptions = new List<Exception>();
+            if (_hostedServices != null) // Started?
+            {
+                foreach (var hostedService in _hostedServices.Reverse())
+                {
+                    try
+                    {
+                        await hostedService.StopAsync(cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (Exception ex)
+                    {
+                        exceptions.Add(ex);
+                    }
+                }
+            }
+
+            // Fire IApplicationLifetime.Stopped
+            _applicationLifetime?.NotifyStopped();
+            await _hostLifetime.StopAsync(cancellationToken);
```

Comment:
Fix this bug, this needs to be stopping, not stopped.

Reply:
IApplicationLifetime.Stopped? IApplicationLifetime.Stopping is triggered before you stop the services, and Stopped is afterwards.

Reply:
Stopping. IHostedService.Stop should map to IApplicationLifetime.Stopping,  it shouldn't be called after.

Reply:
So when does IApplicationLifetime.Stopped get called? or do you plan to eliminate it?
The current flow:
IHost.StopAsync
- IApplicationLifetime.Stopping
- IHostedService(s).StopAsync
- IApplicationLifetime.Stopped
- IHostLifetime.StopAsync

Switching IApplicationLifetime.Stopped and  IHostLifetime.StopAsync might make sense, but Stopping belongs before the calls to Stop.

Reply:
We can consider that, I'm not sure we need Stopped, (I'm not sure why it's used today) Dispose kinda covers it (maybe Stopped maps to Dispose).